database/signals.py

```diff
@@ -0,0 +1,22 @@
+from django.core.exceptions import ValidationError
+from django.dispatch import receiver
+from django.db.models.signals import m2m_changed, post_save
+from .models import Reaction, Stoichiometry
+
+
+def _flatten_stoich_species(stoich_species):
+    return [(stoich, species.id) for stoich, species in stoich_species]
+
+
+@receiver(post_save, sender=Stoichiometry)
+def validate_unique_reaction_stoichiometry(instance, **kwargs):
```

Comment:
Haven't had time to wrap my head around this yet. It's a new feature to me. Maybe you can talk me through it some time. (Just noting here that I haven't reviewed it properly)

Reply:
Yeah this was challenging to find, learn, and implement. I basically wanted to add a constraint that every set of stoichiometric coefficient-species pairs is unique, which should prevent duplicate reactions from getting created. So if we had a reaction like this:
A -> B 
Then no other reaction should have stoichiometry like this: [(-1, A), (1, B)]

It's tricky because it's validating a `ManyToManyField` which has complications in django and the only way to do this is to intercept the signal that gets sent to the database and validate the data there. I need to do a little more checking to make sure this is working.